File: tmf2/org.lttng.scope.tmf2.views.ui/src/org/lttng/scope/tmf2/views/ui/timeline/widgets/timegraph/toolbar/drawnevents/PredicateDrawnEventProvider.java
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.tmf2.views.ui.timeline.widgets.timegraph.toolbar.drawnevents;

import static java.util.Objects.requireNonNull;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.FutureTask;
import java.util.function.Predicate;

import org.eclipse.jdt.annotation.Nullable;
import org.lttng.scope.tmf2.views.core.context.ViewGroupContextManager;

import com.efficios.jabberwocky.common.TimeRange;
import com.efficios.jabberwocky.project.ITraceProject;
import com.efficios.jabberwocky.project.ITraceProjectIterator;
import com.efficios.jabberwocky.timegraph.model.provider.ITimeGraphModelProvider;
import com.efficios.jabberwocky.timegraph.model.provider.drawnevents.TimeGraphDrawnEventProvider;
import com.efficios.jabberwocky.timegraph.model.render.TimeGraphEvent;
import com.efficios.jabberwocky.timegraph.model.render.drawnevents.TimeGraphDrawnEvent;
import com.efficios.jabberwocky.timegraph.model.render.drawnevents.TimeGraphDrawnEventRender;
import com.efficios.jabberwocky.timegraph.model.render.drawnevents.TimeGraphDrawnEventSeries;
import com.efficios.jabberwocky.timegraph.model.render.tree.TimeGraphTreeElement;
import com.efficios.jabberwocky.timegraph.model.render.tree.TimeGraphTreeRender;
import com.efficios.jabberwocky.trace.event.ITraceEvent;
import com.google.common.collect.ImmutableList;

class PredicateDrawnEventProvider extends TimeGraphDrawnEventProvider {

    /** Maximum number of matching events */
    private static final int MAX = 2000;

    private final ITimeGraphModelProvider fModelProvider;
    private final Predicate<ITraceEvent> fPredicate;

    public PredicateDrawnEventProvider(TimeGraphDrawnEventSeries drawnEventSeries,
            ITimeGraphModelProvider modelProvider,
            Predicate<ITraceEvent> predicate) {
        super(drawnEventSeries);
        fModelProvider = modelProvider;
        fPredicate = predicate;

        /* Just use whatever trace is currently active */
        traceProjectProperty().bind(ViewGroupContextManager.getCurrent().currentTraceProjectProperty());
    }

    @Override
    public TimeGraphDrawnEventRender getEventRender(TimeGraphTreeRender treeRender,
            TimeRange timeRange, @Nullable FutureTask<?> task) {

        /*
         * FIXME Extremely slow due to iterating from the start of the trace every
         * single time. JW doesn't allow seeking by timestamp directly yet.
         *
         * Final version should keep an iterator at the last selection, so we can
         * continue using it.
         */
        ITraceProject<?, ?> project = traceProjectProperty().get();
        if (project == null) {
            return new TimeGraphDrawnEventRender(timeRange, Collections.EMPTY_LIST);
        }

        int matches = 0;
        List<ITraceEvent> matchingEvents = new LinkedList<>();
        try (ITraceProjectIterator<?> iter = project.iterator()) {
            while (iter.hasNext()) {
                ITraceEvent event = requireNonNull(iter.next());

                // Replace this by an iterator seek once implemented
                if (event.getTimestamp() < timeRange.getStartTime()) {
                    continue;
                }

                if (event.getTimestamp() > timeRange.getEndTime()) {
                    break;
                }

                if (fPredicate.test(event)) {
                    matchingEvents.add(event);
                    if (matches++ > MAX) {
                        break;
                    }
                }
            }

            // Stream version for Java 9:
//            StreamUtils.getStream(iter)
//                .dropWhile(event -> event.getTimestamp() < timeRange.getStartTime())
//                .takeWhile(event -> event.getTimestamp() <= timeRange.getEndTime())
//                .filter(fPredicate)
//                .limit(MAX)
//                .collect(ImmutableList.toImmutableList());
        }

        List<TimeGraphDrawnEvent> drawnEvents = matchingEvents.stream()
                /* trace event -> TimeGraphEvent */
                .map(traceEvent -> {
                    long timestamp = traceEvent.getTimestamp();
                    /*
                     * Find the matching tree element for this trace event, if
                     * there is one.
                     */
                    Optional<TimeGraphTreeElement> treeElem = treeRender.getAllTreeElements().stream()
                            .filter(elem -> {
                                Predicate<ITraceEvent> predicate = elem.getEventMatching();
                                if (predicate == null) {
                                    return false;
                                }
                                return predicate.test(traceEvent);
                            })
                            .findFirst();

                    if (!treeElem.isPresent()) {
                        return null;
                    }
                    return new TimeGraphEvent(timestamp, treeElem.get());
                })
                .filter(Objects::nonNull)
                /* TimeGraphEvent -> TimeGraphDrawnEvent */
                .map(tgEvent -> new TimeGraphDrawnEvent(tgEvent, getEventSeries(), null))
                .collect(ImmutableList.toImmutableList());

        return new TimeGraphDrawnEventRender(timeRange, drawnEvents);
    }

}

File: lttng-scope/src/main/kotlin/org/lttng/scope/views/timeline/widgets/timegraph/layer/drawnevents/PredicateDrawnEventProvider.kt
/*
 * Copyright (C) 2017 EfficiOS Inc., Alexandre Montplaisir <alexmonthy@efficios.com>
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v1.0 which
 * accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */

package org.lttng.scope.views.timeline.widgets.timegraph.layer.drawnevents

import com.efficios.jabberwocky.common.ConfigOption
import com.efficios.jabberwocky.common.TimeRange
import com.efficios.jabberwocky.context.ViewGroupContext
import com.efficios.jabberwocky.project.TraceProject
import com.efficios.jabberwocky.views.timegraph.model.provider.drawnevents.TimeGraphDrawnEventProvider
import com.efficios.jabberwocky.views.timegraph.model.render.TimeGraphEvent
import com.efficios.jabberwocky.views.timegraph.model.render.drawnevents.TimeGraphDrawnEvent
import com.efficios.jabberwocky.views.timegraph.model.render.drawnevents.TimeGraphDrawnEventRender
import com.efficios.jabberwocky.views.timegraph.model.render.drawnevents.TimeGraphDrawnEventSeries
import com.efficios.jabberwocky.views.timegraph.model.render.tree.TimeGraphTreeRender
import org.lttng.scope.project.filter.EventFilterDefinition
import org.lttng.scope.views.context.ViewGroupContextManager
import java.util.concurrent.FutureTask

/** Maximum number of matching events */
private const val MAX_EVENTS = 2000

/**
 * Provider of drawn event series based on a project event-filter.
 */
class PredicateDrawnEventProvider(private val eventFilter: EventFilterDefinition) : TimeGraphDrawnEventProvider(eventFilter.createSeries()) {

    private val projectChangeListener = object : ViewGroupContext.ProjectChangeListener {
        override fun newProjectCb(newProject: TraceProject<*, *>?) {
            /* Effectively just a bind() */
            traceProject = newProject
        }
    }

    init {
        /* Just use whatever trace is currently active */
        val currentProject = ViewGroupContextManager.getCurrent().registerProjectChangeListener(projectChangeListener)
        traceProject = currentProject

        /* This provider will be active whenever the project filter is enabled */
        enabledProperty().bind(eventFilter.enabledProperty())
    }

    @Suppress("ProtectedInFinal", "Unused")
    protected fun finalize() {
        ViewGroupContextManager.getCurrent().deregisterProjectChangeListener(projectChangeListener)
    }

    override fun getEventRender(treeRender: TimeGraphTreeRender, timeRange: TimeRange, task: FutureTask<*>?): TimeGraphDrawnEventRender {
        val project = traceProject ?: return TimeGraphDrawnEventRender(timeRange, emptyList())

        // TODO We could keep the iterator open for the lifetime of the provider, so that
        // the same iterator is reused from one render to another.

        val matchingEvents = project.iterator().use {
            it.seek(timeRange.startTime)
            it.asSequence()
                    .takeWhile { it.timestamp <= timeRange.endTime }
                    .filter(eventFilter.predicate)
                    .take(MAX_EVENTS)
                    .toList()
        }

        val drawnEvents = matchingEvents
                /* trace event -> TimeGraphEvent */
                .mapNotNull { traceEvent ->
                    treeRender.allTreeElements
                            .find { it.eventMatching?.test(traceEvent) == true }
                            ?.let { TimeGraphEvent(traceEvent.timestamp, it) }
                }
                /* TimeGraphEvent -> TimeGraphDrawnEvent */
                .map { TimeGraphDrawnEvent(it, drawnEventSeries, null) }

        return TimeGraphDrawnEventRender(timeRange, drawnEvents)
    }

}

private fun EventFilterDefinition.createSeries() =
        TimeGraphDrawnEventSeries(this.name,
                ConfigOption(this.color),
                ConfigOption(this.symbol))
